pallets/parachain-staking/src/delegation_requests.rs

```diff
@@ -0,0 +1,330 @@
+						delegator,
+						unstaked_amount: amount,
+					});
+				} else {
+					<DelegationScheduledRequests<T>>::insert(collator, scheduled_requests);
+					<DelegatorState<T>>::insert(&delegator, state);
+				}
+				Ok(().into())
+			}
+			DelegationAction::Decrease(_) => {
+				// remove from pending requests
+				let amount = scheduled_requests.remove(request_idx).action.amount();
+				state.less_total = state.less_total.saturating_sub(amount);
+
+				// decrease delegation
+				for x in &mut state.delegations.0 {
+					if x.owner == collator {
+						return if x.amount > amount {
+							let amount_before: BalanceOf<T> = x.amount.into();
+							x.amount = x.amount.saturating_sub(amount);
+							state.total = state.total.saturating_sub(amount);
+							let new_total: BalanceOf<T> = state.total.into();
+							ensure!(
+								new_total >= T::MinDelegation::get(),
+								<Error<T>>::DelegationBelowMin
+							);
+							ensure!(
+								new_total >= T::MinDelegatorStk::get(),
+								<Error<T>>::DelegatorBondBelowMin
+							);
+							let mut collator_info = <CandidateInfo<T>>::get(&collator)
+								.ok_or(<Error<T>>::CandidateDNE)?;
+							T::Currency::unreserve(&delegator, amount);
+							// need to go into decrease_delegation
+							let in_top = collator_info.decrease_delegation::<T>(
+								&collator,
+								delegator.clone(),
+								amount_before,
+								amount,
+							)?;
+							<CandidateInfo<T>>::insert(&collator, collator_info);
+							let new_total_staked = <Total<T>>::get().saturating_sub(amount);
+							<Total<T>>::put(new_total_staked);
+
+							<DelegationScheduledRequests<T>>::insert(
+								collator.clone(),
+								scheduled_requests,
+							);
+							<DelegatorState<T>>::insert(delegator.clone(), state);
+							Self::deposit_event(Event::DelegationDecreased {
+								delegator,
+								candidate: collator.clone(),
+								amount,
+								in_top,
+							});
+							Ok(().into())
+						} else {
+							// must rm entire delegation if x.amount <= less or cancel request
+							Err(<Error<T>>::DelegationBelowMin.into())
+						};
+					}
+				}
+				Err(<Error<T>>::DelegationDNE.into())
+			}
+		}
+	}
+
+	/// Removes the delegator's existing [ScheduledRequest] towards a given collator.
+	/// Returns [Error::PendingDelegationRequestDNE] if request does not exist.
+	pub(crate) fn delegation_remove_request_with_state(
+		collator: &T::AccountId,
+		delegator: &T::AccountId,
+		state: &mut Delegator<T::AccountId, BalanceOf<T>>,
+	) -> DispatchResultWithPostInfo {
+		let mut scheduled_requests = <DelegationScheduledRequests<T>>::get(collator);
+
+		let request_idx = scheduled_requests
+			.iter()
+			.position(|x| &x.delegator == delegator)
+			.ok_or(<Error<T>>::PendingDelegationRequestDNE)?;
+
+		let request = scheduled_requests.remove(request_idx);
+		let amount = request.action.amount();
+		state.less_total = state.less_total.saturating_sub(amount);
```

Comment:
note to self: verify that caller of this function writes `state` to storage after calling (we should add a comment above the function to ensure this is done in the future as well)

alternatively could write to storage and if it is rewritten then changes are cached. This would be safer
```suggestion
		state.less_total = state.less_total.saturating_sub(amount);
		<DelegatorState<T>>::insert(delegator, state)
```

Reply:
Yeah I thought about it. So far adding a comment could be the bare minimum we should do.

The problem with writing to state here is that the caller of this function is usually operating on the `mut state` (like [here](https://github.com/PureStake/moonbeam/pull/1408/files/5ca3822697cb279e58a8e6c99e318068ce3a61e9#diff-842251b834245800808bd740184c1e3b18c6af1f95fb1bd4d484420017fdf86eL989)) and as such if we write the storage here itself, then the previously read value becomes dirty in the caller function - which can actually cause more bugs.

Reply:
> We however cannot guarantee that the caller never cloned the state and used it elsewhere. So that should still be taken into account.

Good point